pre_commit/languages/download.py

```diff
@@ -0,0 +1,258 @@
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.envcontext import Var
+from pre_commit.prefix import Prefix
+
+ENVIRONMENT_DIR = 'download'
+get_default_version = lang_base.basic_get_default_version
+run_hook = lang_base.basic_run_hook
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix, version: str) -> Generator[None, None, None]:
+    envdir = lang_base.environment_dir(prefix, ENVIRONMENT_DIR, version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+@dataclass(frozen=True)
+class Platform:
+    value: str
+
+    def __post_init__(self) -> None:
+        os, cpu = self.parts
+
+        if os not in {'linux', 'darwin', 'windows'}:
+            raise ValueError(f'invalid operating system `{os}`')
+
+        if cpu not in {'amd64', 'aarch64'}:
+            raise ValueError(f'invalid CPU `{cpu}`')
+
+    @property
+    def parts(self) -> tuple[str, str]:
+        first, second = self.value.split('/', 1)
+        return (first, second)
+
+    @property
+    def os(self) -> Literal[
+        'linux', 'darwin', 'windows', 'dragonfly',
+        'freebsd',
+    ]:
+        os, _ = self.parts
+        return os  # type: ignore
+
+    @property
+    def cpu(self) -> Literal[
+        'arm64', 'arm64be', 'arm', '386', 'amd64', 'ppc',
+        'ppc64', 'ppc64le',
+    ]:
+        _, cpu = self.parts
+        return cpu  # type: ignore
+
+    @staticmethod
+    def host() -> Platform:
+        os = {
+            'Linux': 'linux',
+            'Darwin': 'darwin',
+            'Windows': 'windows',
+            'DragonFly': 'dragonfly',
+            'FreeBSD': 'freebsd',
+        }[platform.system()]
+        cpu = {
+            'aarch64': 'arm64',
+            'aarch64_be': 'arm64be',
+            'arm': 'arm',
+            'i386': '386',
+            'i686': '386',
+            'x86_64': 'amd64',
+            'ppc': 'ppc',
+            'ppc64': 'ppc64',
+            'ppc64le': 'ppc64le',
+        }[platform.machine()]
+        return Platform(f'{os}/{cpu}')
+
+
+class ContentCorruptionError(Exception):
+    pass
+
+
+class SRI:
+    def __init__(self, value: str):
+        self._value = value
+        self._algorithm, self._checksum = self.value.split('-', 1)
+        assert self.algorithm in {'sha256', 'sha384', 'sha512'}
+
+        if b64encode(b64decode(self.checksum)).decode('utf-8') !=\
+                self.checksum:
+            raise ValueError('Invalid checksum string.')
+        if (self.algorithm == 'sha256' and len(self.checksum) != 44) or \
+           (self.algorithm == 'sha384' and len(self.checksum) != 64) or \
+           (self.algorithm == 'sha384' and len(self.checksum) != 88):
+            raise ValueError(
+                f'Invalid checksum string length for {self.algorithm}',
+            )
```

Comment:
Use [digest_size](https://docs.python.org/3/library/hashlib.html#hashlib.hash.digest_size).

```suggestion
        hasher = hashlib.new(self.algorithm)
        if self.checksum != hasher.digest_size:
            ...
```

Reply:
Done